lib/tests/test_git.rs

```diff
@@ -2781,6 +2797,381 @@ fn test_push_branches_not_fast_forward_with_force() {
 }
 
+fn assert_resolves_to_self(r: &RefTarget, merge_fun: impl Fn(&RefTarget) -> RefTarget) {
+    assert_eq!(r, &merge_fun(r));
+}
+fn assert_fails_to_resolve_to_self(
+    r: &RefTarget,
+    merge_fun: impl Fn(&RefTarget) -> RefTarget,
+) -> Merge<Option<CommitId>> {
+    let result = merge_fun(r);
+    assert_ne!(r, &result);
+    result.as_merge().clone()
+}
+
+// TODO(ilyagr): More tests for push safety checks were originally planned. We
+// may want to add tests for when a branch unexpectedly moved backwards or
+// unexpectedly does not exist for branch deletion.
+
+#[test]
+fn test_push_updates_unexpectedly_moved_sideways_on_remote() {
+    let settings = testutils::user_settings();
+    let temp_dir = testutils::new_temp_dir();
+    let setup = set_up_push_repos(&settings, &temp_dir);
+    let index = setup.jj_repo.index();
+
+    // The main branch is actually at `main_commit` on the remote. If we expect
+    // it to be at `sideways_commit`, it unexpectedly moved sideways from our
+    // perspective.
+    //
+    // We cannot delete it or move it anywhere else. However, "moving" it to the
+    // same place it already is is OK, following the behavior in
+    // `test_merge_ref_targets`.
+    //
+    // For each test, we check that the push succeeds if and only if the branch
+    // conflict `jj git fetch` would generate resolves to the push destination.
+
+    let merge_with_main_commit_and_sideways_as_base = |r: &RefTarget| {
+        merge_ref_targets(
+            index,
+            r,
+            &RefTarget::normal(setup.sideways_commit.id().clone()),
+            &RefTarget::normal(setup.main_commit.id().clone()),
+        )
+    };
+    let attempt_push_expecting_sideways = |target: Option<CommitId>| {
+        let targets = [GitRefUpdate {
+            qualified_name: "refs/heads/main".to_string(),
+            force: true,
+            expected_current_target: Some(setup.sideways_commit.id().clone()),
+            new_target: target,
+        }];
+        git::push_updates(
+            &get_git_repo(&setup.jj_repo),
+            "origin",
+            &targets,
+            git::RemoteCallbacks::default(),
+        )
+    };
+
+    let insta_settings = setup.add_commit_alias_filters(insta::Settings::clone_current());
+    insta_settings.bind(||{
+    assert_debug_snapshot!(
+      assert_fails_to_resolve_to_self(
+        RefTarget::absent_ref(),
+        merge_with_main_commit_and_sideways_as_base
+      ),
+      @r###"
+    Conflicted(
+        [
+            None,
+            Some(
+                CommitId(
+                    "sideways",
+                ),
+            ),
+            Some(
+                CommitId(
+                    "main",
+                ),
+            ),
+        ],
+    )
+    "###
+    );
+    let result = attempt_push_expecting_sideways(None);
+    assert_snapshot!(format!("{result:?}"), @r###"Err(RefInUnexpectedLocation(["refs/heads/main"]))"###);
```

Comment:
nit: it can be `insta::assert_debug_snapshot!()`, but I would probably use `assert_matches!(.., Err(RefInUnexpectedLocation(_)))` since the failed ref names aren't important.